### ISSUE TITLE
Effectful imports should ignore sideEffects field

### ISSUE THREAD

**Issue by Andarist** [webpack/webpack#6571]
**Do you want to request a *feature* or report a *bug*?**

feature

**What is the current behavior?**

```
import 'list/fantasy-land'
```
with 
```js
sideEffects: false
```

does not include imported file in the bundle, which is somewhat OK

**If the current behavior is a bug, please provide the steps to reproduce.**

not a bug

**What is the expected behavior?**

this import is **clearly** effectful, so it actually IMHO should not get dropped regardless of the `sideEffects`

**If this is a feature request, what is motivation or use case for changing the behavior?**

described in expected behaviour

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

webpack@4, others - irrelevant

**Comment by sokra** [comment]
If `import 'list/fantasy-land'` has a side-effect, you `sideEffects` configuration is wrong.

You can and should include `list/fantasy-land` in your configuration like this:

``` js
// package.json of 'list'
{
  "sideEffects": [
    "./fantasy-land"
  ]
}

**Comment by Andarist** [comment]
I understand how to solve it, [I've did it](https://github.com/funkia/list/pull/17/files) just before posting an issue here and I understand this is not the most popular case, but still think that my argument has some merit - `import 'package'` is side-effectful "by definition" and `sideEffects` could skip such imports by default.

**Comment by probablyup** [comment]
I agree with @Andarist. If you consider the import statement, a naked import implies “execute everything in this file”, which is side effectful by default.

**Comment by dcporter** [comment]
I agree with @Andarist. [MDN describes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) `import '/modules/my-module.js';` as "Import a module for its side effects only", and such imports have no other purpose. The use of this form of import syntax is sufficient information to indicate that the imported module has side effects.

So the developer has already specified that this module has side effects; the requirement to specify it *again* in a separate location is suboptimal and fragile.

**Comment by rtsao** [comment]
In practice, it might be nice to configure webpack to assume all modules are side effect free by default, as the whole npm ecosystem doesn't use it yet.

If webpack treated an import statement with no specifiers (i.e. `import "foo"`) as a signal for side effects, this would be a convenient, zero-config way to opt out of this behavior.

For example:

```js
import {something} from "foo";

// unused import gets pruned
```

```js

import "foo";
// implies side effects

import {something} from "foo";

// unused import does not get pruned

```

This behavior could be nice combination with:
- Warnings when imports are pruned but not explicitly `"sideEffects": false`
- Warnings when imports are explicitly `"sideEffects": false` but they have been imported without specifiers (i.e. `import "foo";`)

**Comment by edmorley** [comment]
Hi

In facebook/create-react-app#5140 the package in question mistakenly set `"sideEffects": false` even though it includes CSS, which caused the CSS to be dropped when `mode` was `'production'`.

This lead to the discussion here:
https://github.com/facebook/create-react-app/issues/5141#issuecomment-425276343

@sokra, what do you think about the idea of allowing loaders to declare that their output always has side effects, thereby overriding the package.json `sideEffects` (perhaps emitting a warning whilst doing so)? css-loader could then use this to declare that its output should never be optimized away by `SideEffectsFlagPlugin`.

**Comment by gaearon** [comment]
I disagree that the import form should determine dead code elimination. That sounds like a dangerous heuristic to me because it’s so subtle. I think it’s fair that the package author should declare effectful entry points separately. However I also think webpack could *warn* when you use such an entry point with a plain import.

That said the way webpack handles *CSS* in this case doesn’t make sense to me. See https://github.com/facebook/create-react-app/issues/5141#issuecomment-425290508. When you import CSS package library can’t “know” how you’ll “interpret” it. So it can’t know whether it will become a side effect or not. The real effect here is in the style loader. So that leads me to thinking loaders should be able to mark their output as effectful.

**Comment by rtsao** [comment]
It would be nice if the entire npm ecosystem adopted the sideEffects package.json field, but realistically, I think it will be a long time before that is the case.

Consider the following scenario:

- The top-level configuration of webpack has been set to`sideEffects: false`.
- Package "core-js" is explicitly imported for side effects (i.e. `import "core-js";`) *and* omits a `sideEffects` field in its package.json

In this scenario, I think pruning "core-js" is clearly incorrect behavior. I certainly don't think import form should be the *only* heuristic for determining dead code elimination, but in this specific case where a package's side effectfulness is not explicitly specified, not using that information seems just plain wrong.

Until a majority of the npm ecosystem specifies a `sideEffects` field, it would be nice to allow users to have a default behavior of pruning unused imports (which corresponds to the vast majority of packages) but allow unambiguously effectful imports (`import "foo";`) to be preserved without extra configuration.

**Comment by gaearon** [comment]
@rtsao I think in the scenario you're describing treating such imports specially only adds to the overall confusion. Because now some cases will work and some wouldn't, but the problem will get less widespread and thus there'll be less incentives to properly fix it. So it could cause more pain in longer term. Your solution is probably reasonable for the behavior when **the user explicitly sets top-level `sideEffects` to `false`** (which I didn't even know is possible — how can the user know everything about their dependencies? this sounds like a dangerous option that most people will misunderstand). But for cases when you don't have anything on the top level, I don't think plain `import` should be treated as effectful **in contrast to package's own explicit declaration**. For those cases IMO webpack should warn but still respect the package declaration. Otherwise it's too unpredictable.

**Comment by dcporter** [comment]
The `import “foo”` form is an extremely clear indication of developer intent. There is no case in which one uses that import form and *doesn’t* want it included (assuming the importing file is included) — that’s what that form is for, and it’s the only thing it’s for. That clear, language-level developer declaration should in *all* cases override the sideEffects flag, and not just for now—broad, clear understanding and use of the flag won’t change the clarity of developer intent when importing a module without symbol.

**Comment by gaearon** [comment]
The reason I'm opposed to this is because it's unexpected to see refactoring from

```js
import 'foo'
```

to 

```js
import {x} from 'foo'
```

break code.

For example, somebody *might* be running some code just for side effects, relying on your proposed behavior. Then they actually need to add an import, and it's super easy to miss that *adding* an import would *remove* the side effects *in production only*.

Of course you can have two `import` statements but then *that* is unexpected too.

```js
import 'foo'
import {x} from 'foo'
```

Anybody can refactor this to "remove duplicate import" (in fact there's probably linter plugins to do that or complain about that). It's too subtle for this to change the behavior.

This is why I think webpack should *warn* when you use plain import (correctly guessing your intent, and surfacing the problem to you). But I don't think it should change the behavior. Because the fix should be in the package that "lies" about its declared side effects.

**Comment by sokra** [comment]
A warning is definitely a good idea.

Using `import "x"` when x is sideeffect-free seems to be wrong to me, either x isn't really sideeffect-free and the package.json is incorrect or the import should be removed because it has no effect. So accepting `import "x"` as forced inclusion doesn't seem to be a valid option to me.

**Comment by sokra** [comment]
Some note about adding `sideEffects: false` to your application. This will not inherit to the node_modules. Only the nearest package.json in the tree is used to read the flag. -> flag affects only the package and not dependencies

**Comment by sokra** [comment]
Regarding importing CSS. While the exact way how CSS is handled is maybe not defined, the semantic of `import "file.css"` is defined. It means: "the rules in the file become somehow effective". And the semantic is also the thing that defines if it has sideeffects, which is has. So a module can flag CSS as sideeffect-full even if it doesn't know how it's handled exactly.

Anyway, you can also set sideEffects in the CSS rule in `module.rules`>

**Comment by gaearon** [comment]
Thanks, this makes sense!

**Comment by gaearon** [comment]
>Anyway, you can also set sideEffects in the CSS rule in module.rules.

Do you feel it would be a reasonable default though? We need to decide what CRA should do in https://github.com/facebook/create-react-app/issues/5140.

**Comment by dcporter** [comment]
> Using import "x" when x is sideeffect-free seems to be wrong to me

If the developer wrote `import “x”`, the developer said to include it for its side effects, full stop. Webpack shouldn’t be thinking any more about it than that. (Again, assuming that the importing file is included.) If the developer got that wrong, that’s fine, they get an extra file that they explicitly asked for. Again, accepting it as forced inclusion is what that form exists for, and is the only thing it exists for.

> it's unexpected to see refactoring from
>
> `import 'foo'`
>
> to
>
> `import {x} from 'foo'`
>
> break code.

I agree that’s a surprising result, and I agree that having two imports is gross (and probably lint-breaking). But the refactored code is incorrect (and is therefore not a refactor): importing `x` *and not using it* is a mistake, and isn’t the same as importing `foo` for side effects. I’d also expect it to be a rarely-sprung trap: most side effect files (e.g. CSS files) have no exports. So breaking a language-level intuition to protect developers from the results of a rare, double mistake, feels like a pretty bad trade-off to me.

Anecdotally, every coworker whose opinion I have initially expects side-effect imports to work with `sideEffects: false`. (Similarly unscientific: my post above has ten 👍s.) The idea that the side-effect import should work is a simple *language-level* intuition, which is correct inasmuch as it’s backed up by MDN. `sideEffects` should honor it, and can do so without being less useful or powerful.

**Comment by sokra** [comment]
> Do you feel it would be a reasonable default though?

Sideeffect-full is the default. You opt-in into sideeffect-free with the package.json-flag.

Being always sideeffect-full for CSS feels like not respecting the opt-in in the package.json and promoting an incorrect sideEffects config in package.json.

Note that CSS modules can be flagged as sideeffect-free.

**Comment by sokra** [comment]
> If the developer wrote import “x”, the developer said to include it for its side effects, full stop.

If the developer wrote `"sideEffects": false`, the developer said that it has no side effects, full stop.

These are two conflicting statements the developer made here. Maybe we should even emit an error here.

**Comment by dcporter** [comment]
> Anecdotally, every coworker whose opinion I have initially expects side-effect imports to work with `sideEffects: false`.

(These are folks who get what the flag is for and think it’s awesome, by the way.)

**Comment by gaearon** [comment]
>While the exact way how CSS is handled is maybe not defined, the semantic of import "file.css" is defined. It means: "the rules in the file become somehow effective". And the semantic is also the thing that defines if it has sideeffects, which is has. So a module can flag CSS as sideeffect-full even if it doesn't know how it's handled exactly.

Based on this, it seems like webpack could always error if you import CSS from package that says it has no side effects? Since it's never correct.

**Comment by gaearon** [comment]
In Create React App, we've decided to always treat css/sass/scss as side effectful in the module rules. If webpack adds a warning (or better, an error) when CSS is marked as side effect-less, we can remove this customization. But for now it's too subtle and would bite people in production otherwise.

**Comment by satya164** [comment]
@dcporter did you read Dan's comment? where he gives this example:

The developer imports a module for side-effects:

```js
import 'some-module';
```

Later they want to import something else from the package:

```js
import { x } from 'some-module';
```

Now, after this change, only `x` is going to be imported and the code doing the side-effect might be tree-shaked and dropped, breaking the code. Webpack can't know and fix this, because the package author lied about the package being side-effect free in the `sideEffects: false` field. So the heuristics can introduce subtle bugs when you refactor code.

Webpack could apply the heuristic that a module has no side-effect when the user does `import 'some-module';`, but here the question is, who does it trust? The user or the author, both have conflicting opinions about the package being side-effect free.

**Comment by dcporter** [comment]
@satya164 yes I did, did you read [my response](
https://github.com/webpack/webpack/issues/6571#issuecomment-425668670)? I don't believe webpack does sub-module tree shaking, so the only way that this would trigger a bug is if you import { x } but don't use it; most side-effect files (especially but not only CSS files) will never go from not having exports to having exports; so this "refactor" is almost certainly a mistake, and doubly rare. In a competition for least-likely-to-cause-weird-breaks, I think that wins handily over the `sideEffect: false` status quo, which has surprised every team I'm aware of when they start using it, and sets many teams up for a fragile future if they ever add any not-easily-regexable side effect files (polyfills or whatever).

> the package author lied about the package being side-effect free in the `sideEffects: false` field.

You're right. But the whole point of this Issue is about changing the heuristics of `sideEffects: false` to define the behavior this currently-undefined conflict.

> the question is, who does it trust?

Again, my contention is that given two conflicting instructions here, the one that's a language feature that can only possibly mean "import this anyway" should take precedent. I'm not asking for Webpack to get all weird and heuristicy, I'm saying that the *intended meaning* of `sideEffects: false` could change for the better.

Good point about importing a whole package this way — I'm not sure how this should play out. Probably the change we're requesting should be minimal, and only apply down one layer. (So the imported package's index file would be imported, but non-side-effectful re-exports from *that* file could be stripped, meaning only the package's own first-tier side-effect-y imports would be imported.)

@gaearon to answer your question, flagging the CSS files as having side effects in the `package.json` is exactly the right thing to do today. I'd love to know if you have any thoughts about my response to your downvote of this Issue's requested change.

**Comment by edmorley** [comment]
Searching GitHub issues for "webpack css sideEffects" shows there appears to be pretty widespread confusion regarding `"sideEffects"` and CSS with the current API design/docs/lack of warning. 

For example in addition to the issues linked from https://github.com/webpack/webpack/issues/6571#issuecomment-425366992 (and only following the first N pages of results) I found:
- https://github.com/webpack/webpack/issues/7094#issuecomment-419823040
- https://github.com/webpack/webpack/issues/7491
- https://github.com/webpack-contrib/mini-css-extract-plugin/issues/102
- https://github.com/webpack-contrib/mini-css-extract-plugin/issues/118
- https://github.com/webpack-contrib/mini-css-extract-plugin/issues/245
- https://github.com/webpack-contrib/mini-css-extract-plugin/issues/275
- https://github.com/webpack-contrib/sass-loader/issues/553
- https://github.com/webpack-contrib/style-loader/issues/312
- https://github.com/ant-design/ant-design/issues/10190
- https://github.com/bvaughn/react-virtualized/issues/1135
- https://github.com/CompuIves/codesandbox-client/pull/1133
- https://github.com/FormidableLabs/react-live/pull/74
- https://github.com/Microsoft/monaco-editor/pull/981#issuecomment-412510348
- https://github.com/vuejs/vue-cli/issues/1287
- https://github.com/vuejs/vuepress/issues/718

**Comment by jasonk** [comment]
FWIW, I agree with the sentiment that `import "./file";` should work this way, largely because while the `sideEffects` flag is useful, it's also terribly confusing.

For example here are snippets of some of the files from the project where I ran into this issue:

```javascript
// entry.js
import './icons';
// ..bunch of other irrelevant imports

// icons/index.js
import './library';
export * from './Icon';

// icons/library.js
import { library } from '@fortawesome/fontawesome-svg-core';
import * as icons from './icon-registry';
library.add( ...Object.values( icons ) );

// icons/icon-registry.js
export {
  faUserCircle as farUserCircle,
  faCircle as farCircle,
  faIdCard as farIdCard,
} from '@fortawesome/free-regular-svg-icons';
```

The problem that I had is that when running `webpack --mode=production` none of the icons made it into the bundle, but it worked fine when running with `mode=development`.  I had read through the tree-shaking docs and knew about `sideEffects` so I added the module that had side effects to my `package.json`:

```json
"sideEffects": [ "src/icons/library.js" ],
```

This didn't do anything.  After some experimentation I arrived at:

```json
"sideEffects": [ "src/icons/library.js", "src/icons/index.js" ],
```

So basically it seems that part of the issue with `sideEffects` is that while you clearly have to include modules that have actual side-effects, in some situations you also need to include modules that simply import things that have side effects.  Or maybe the problem is that "side effects" is not well-defined.  In either case, having `import "./foo"` be treated as having side effects would avoid this confusion (and given the number of posts I found from people having problems with this, it seems to be a fairly common source of confusion).

**Comment by Leedehai** [comment]
I respectfully disagree with this feature request.

To follow the [Principle of Least Astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), if a developer wish to keep a `import 'foo.js';` effective in production build, they should declare `foo.js` to have side effects in package.json's 'sideEffects' list. If a file is missing from this whitelist, Webpack should happily prune the file away in production mode.

In a practical use, projects may be using override that should only be effective in development mode (e.g. for test mocking):
```
// some-module.js
import 'some-override'

// ...
```

And developers should be confident this polyfill won't be included in production build.

**Comment by dcporter** [comment]
I agree wholeheartedly with the principle of least astonishment. I'm astonished that anyone considers "imports should work the way that the _JavaScript language says they should work_ " more astonishing than "a `package.json` flag that webpack made up a year and a half ago should override the language".

> In a practical use, projects may be using overrides that **should only be effective in dev mode**

I'm not clear what you're describing here, does `sideEffects` behave differently in prod than in dev? That's terrifying. Changing import behavior between dev and prod is a deeply hacky, wrong-layer use of this `package.json` feature. That sort of behavior should live in webpack configs, and should definitely _not_ live in `package.json`. `import` behavior should absolutely not _implicitly_ change between dev and prod like you describe, that's a massive trap.

> Adding implicit deductions in this decision process, as proposed by this feature request, **does not scale out well**

JavaScript has a clear language-level way to declare side-effectful imports. This proposal doesn't suggest adding anything implicit: it asks that this flag from a _build tool_ not override the language. Developers know what `import` means, and have never heard of `sideEffects` until they're _deep_ into webpack academia.

I would ask that the webpack team considers what's expected, surprising, and explicit from the perspective of developers using the JavaScript language itself. Every single time we've used this feature with new teams, developers are astonished that their clear-as-day side-effectful imports (`import "./path/to/file.css"`) get stripped.

(sidebar: <3 your avatar.)

**Comment by chybisov** [comment]
Any updates on this?

**Comment by geoffp** [comment]
Just found this because I am _astonished_ that Webpack is throwing away all my hard work on a polyfill loader. :)

FWIW, I also expected that `import "./sideEffects";` would not be tree-shaken out by default, since that is, semantically, an explicit import of something that can only have side effects. I've gotten it working now, but only after a long struggle.

The documentation on `sideEffects` could probably use a little detail on what's necessary for library authors, and also make clear that you have to list "import ancestors" of side-effecty files you want to keep. I found that I had to both list ancestors, and make the paths point at build artifacts (in `./dist/`, in my case) rather than source files. Like so:
```json
"sideEffects": [
  "./dist/polyfills/index.js",
  "./dist/index.js"
]
```
(Assuming that `index.js` imports `./polyfills`). The first line on its own appeared not to be enough, as others have mentioned.

**Comment by strass** [comment]
I agree that this is a cruel trap and that it should at least be consistent between dev and prod

**Comment by scamden** [comment]
@dcporter 
>  I don't believe webpack does sub-module tree shaking, 

Are you sure about that? Perhaps it is due to two different optimizations (usedExports vs sideEffects) but if I add an exported function to a module and don't import it webpack definitely drops it.

**Comment by sokra** [comment]
> I agree that this is a cruel trap and that it should at least be consistent between dev and prod

In webpack 5 the side effects optimization is enabled for both modes to reduce differences.

**Comment by scamden** [comment]
@sokra do the marked side effects also get dropped even without minification?

**Comment by sokra** [comment]
Yes

**Comment by applemonkey496** [comment]
Wait, so was this feature actually implemented? Is it safe to use:
 ```ts
import 'file.css'
// or
import 'file.js'

// With { sideEffects: false } in package.json?
// -------------
// Will the contents of file.js appear in the bundle?
```

**Comment by Andarist** [comment]
> Will the contents of file.js appear in the bundle?

No, since you have configured `sideEffects: false` and no import from that file ends up being used.

**Comment by benjamingr** [comment]
Like @gaearon this seems like it should be up to the loader or plugins shouldn't it?

Is there a way to explicitly mark a module as "having side effects" from a loader/plugin currently?

**Comment by sachinlala** [comment]
Is there a new update on this or a solution ?

**Comment by olsonpm** [comment]
I disagree the behavior of webpack should differ from a module's configured "sideEffect" flag.  Implicit overrides to an explicit configuration would make things even more confusing than they already are.

**Comment by dcporter** [comment]
@alexander-akait great news, curious when/where this was this completed?

**Comment by alexander-akait** [comment]
@dcporter What we should to solve here? Answer is here - https://github.com/webpack/webpack/issues/6571#issuecomment-368405434

**Comment by dcporter** [comment]
_scrolls down_ You mean way up above the eighty other subsequent comments? What should be solved is webpack considering its made-up config to be superior to most clear possible statement of intent from the JavaScript language itself.

**Comment by alfaproject** [comment]
Or at least opt-in to something implicit to keep those imports:
```
{
  "sideEffects": "implicit"
}
```

**Comment by alexander-akait** [comment]
@dcporter We use such approach a lot of time (since 2018), changing it right now will be very destructive (a lot of other tools use the same approach) and a lot of project will lose this optimization (bigger size). 

By default we use this https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L1464 for webpack@5, i.e. webpack respects `sideEffect` in `package.json` for dev and prod env and you will not faced with different output. If you set `sideEffects: false`, you clearly say - all my imports are side effect. `sideEffects` is special thing for bundling where you should clearly indicate where there is a side effects where there is none.

I am not against implementing hints or warnings about potential places, but no one has tried to implement it so far, so I think it is not so popular, and therefore closed the issue

Where and which do you have a problem? 

@alfaproject 

> Or at least opt-in to something implicit to keep those imports:

This idea is good, but we have vite/esbuild/rspack/etc and they don't support this value, integrate it can be very long and I don't library authors really can use it someday (don't forget about enterprise that may use the old version for a long time).

I don't think this is a problem specifically with webpack now, although it was originally implemented by webpack, change the behavior or add new values will require coordination of all the authors of the tools, a long time of for implementation, releases, updating, and then a long time for authors.

**Comment by alexander-akait** [comment]
> Approaching 100% of greenfield projects set up by non webpack experts have this real problem. (It's not as atrocious as it used to be, because now you have to fix it up front instead of discovering it just before you ship.)

What projects? If you don't have `sideEffect` you will not have problems, but if you use it, please read docs how it works

> I disagree that it would be unduly disruptive to fix this. The deoptimization would be rare since most extant projects have been forced to work through it.

I disagree too, it is not rare and I faced with it a lot of time, and a lot of time in big projects, I suggest you provide constructive arguments, rather than stating who agrees or disagrees, this is counter-constructive.

> Other bundlers have implemented side-effect flags whose defaults ignore JavaScript semantics? Which ones?

Which not? https://github.com/evanw/esbuild/issues/1241#issuecomment-831130652 

I can provide more link if you want

Once there was a decision to implement this solution like this, yes it may not be perfect and there are certain gaps in the design, but it was done like this, and now changing it will cause huge problems in the entire ecosystem, and I'd rather not touch it, if we want to improve/change/make it more flexible we will need a new field with new structures and values (for example we can take a `sideEffectsMapping` name where you can provide more values, like Node.js makes with `main` and `exports` fields). 

If you want to solve this, I suggest you create an RFC, describe your vision and invite the main developers of the tools for discussions, after this is approved we can begin implementation, this is how backward compatibility works and this is how the process in the ecosystem occurs, I can't just change something if it's been working for years just because someone wanted it, sorry.
